In [1]:
dataset_path = r"C:\Users\wwwad\.cache\kagglehub\datasets\msambare\fer2013\versions\1"

In [2]:
import kagglehub 
path = kagglehub.dataset_download("msambare/fer2013")
print("Path to dataset: " , path)

c:\Users\wwwad\tf_directml310\lib\site-packages\tqdm\auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


Path to dataset:  C:\Users\wwwad\.cache\kagglehub\datasets\msambare\fer2013\versions\1


In [2]:
import tensorflow as tf
print(tf.config.list_physical_devices('GPU'))

[PhysicalDevice(name='/physical_device:GPU:0', device_type='GPU'), PhysicalDevice(name='/physical_device:GPU:1', device_type='GPU')]


In [3]:
import sys 
print(sys.executable)

c:\Users\wwwad\tf_directml310\Scripts\python.exe


In [4]:
import os 
print(os.listdir(dataset_path))

['test', 'train']


In [5]:
train_dataset_path = os.path.join(dataset_path , "train")
test_dataset_path = os.path.join(dataset_path , "test")

In [6]:
print(os.listdir(train_dataset_path))
print(os.listdir(test_dataset_path))

['angry', 'disgust', 'fear', 'happy', 'neutral', 'sad', 'surprise']
['angry', 'disgust', 'fear', 'happy', 'neutral', 'sad', 'surprise']


In [7]:
from tensorflow.keras.preprocessing.image import ImageDataGenerator 
train_gen = ImageDataGenerator(rescale = 1./255,  rotation_range = 15 , zoom_range = 0.2 , horizontal_flip = True , validation_split = 0.25 , brightness_range = [0.8 , 1.2])
val_gen = ImageDataGenerator(rescale = 1./255 , validation_split = 0.25)
train_generator = train_gen.flow_from_directory(train_dataset_path , target_size = (224 , 224) , batch_size = 32 ,  color_mode = "rgb" ,  subset = "training" , class_mode = "categorical" , seed = 42 )
Val_generator = val_gen.flow_from_directory(train_dataset_path , target_size = (224 , 224) , batch_size = 32 , color_mode = "rgb" , subset = "validation"  , class_mode = "categorical" , seed = 42)

Found 21535 images belonging to 7 classes.
Found 7174 images belonging to 7 classes.


In [8]:
from collections import Counter
Counter(train_generator.classes)

Counter({3: 5412, 4: 3724, 5: 3623, 2: 3073, 0: 2997, 6: 2379, 1: 327})

In [62]:
history.history['accuracy']
history.history['val_accuracy']

[0.3797045052051544,
 0.48912742733955383,
 0.5425146818161011,
 0.5539448261260986,
 0.5761082172393799,
 0.5459994673728943,
 0.5954837203025818,
 0.5517145395278931,
 0.561890184879303,
 0.6032896637916565,
 0.5904656052589417,
 0.6293560266494751,
 0.6055199503898621,
 0.5965988636016846,
 0.6191803812980652,
 0.6414831876754761,
 0.6204349398612976,
 0.6201561689376831,
 0.6003624796867371,
 0.6346529722213745,
 0.6253136396408081]

In [60]:
train_generator.classes

array([0, 0, 0, ..., 6, 6, 6])

In [26]:

test_gen = ImageDataGenerator(rescale = 1./255)
test_generator = test_gen.flow_from_directory(test_dataset_path , batch_size = 32 , target_size=(224,224) , shuffle = False)

Found 7178 images belonging to 7 classes.


In [31]:
len(Labels_test)

NameError: name 'Labels_test' is not defined

In [141]:
len(Image_test)

38

In [143]:
Image_test = np.array(Image_test)

In [8]:
import tensorflow 
from tensorflow import keras 
from keras import Sequential
from keras.layers import Dense , Conv2D , Input , MaxPooling2D , AveragePooling2D , GlobalAveragePooling2D , Flatten

## Model training part---------->

In [185]:
model = Sequential()
model.add(Conv2D(32 , kernel_size = (2 , 2) ,   padding = "same" , activation = "relu" , input_shape = (224 , 224 , 3)))
model.add(Conv2D(32 , kernel_size = (2 , 2) ,   padding = "same" , activation = "relu"))
model.add(MaxPooling2D(pool_size = (2 , 2)))

model.add(Conv2D(64 , kernel_size = (3 , 3) , activation = "relu"  , padding = "same"))
model.add(Conv2D(64 , kernel_size = (3 , 3) ,   padding = "same" , activation = "relu"))
model.add(MaxPooling2D(pool_size = (2 , 2)))



model.add(GlobalAveragePooling2D())

model.add(Dense(108 , activation = "relu"))
model.add(Dense(7, activation = "softmax"))

Epoch 1/30
673/673 [==============================] - 182s 270ms/step - loss: 1.8083 - accuracy: 0.2491 - val_loss: 1.7904 - val_accuracy: 0.2541
Epoch 2/30
673/673 [==============================] - 146s 217ms/step - loss: 1.7917 - accuracy: 0.2504 - val_loss: 1.7887 - val_accuracy: 0.2537
Epoch 3/30
673/673 [==============================] - 148s 219ms/step - loss: 1.7842 - accuracy: 0.2547 - val_loss: 1.7818 - val_accuracy: 0.2584
Epoch 4/30
673/673 [==============================] - 156s 232ms/step - loss: 1.7760 - accuracy: 0.2590 - val_loss: 1.7776 - val_accuracy: 0.2658
Epoch 5/30
673/673 [==============================] - 139s 207ms/step - loss: 1.7687 - accuracy: 0.2615 - val_loss: 1.7649 - val_accuracy: 0.2665
Epoch 6/30
673/673 [==============================] - 217s 322ms/step - loss: 1.7644 - accuracy: 0.2707 - val_loss: 1.7634 - val_accuracy: 0.2763
Epoch 7/30
673/673 [==============================] - 220s 327ms/step - loss: 1.7589 - accuracy: 0.2715 - val_loss: 1.7671 - val_accuracy: 0.2752
Epoch 8/30
673/673 [==============================] - 219s 326ms/step - loss: 1.7536 - accuracy: 0.2782 - val_loss: 1.7609 - val_accuracy: 0.2792
Epoch 9/30
673/673 [==============================] - 202s 299ms/step - loss: 1.7491 - accuracy: 0.2834 - val_loss: 1.7453 - val_accuracy: 0.2901
Epoch 10/30
673/673 [==============================] - 143s 212ms/step - loss: 1.7425 - accuracy: 0.2873 - val_loss: 1.7436 - val_accuracy: 0.2957
Epoch 11/30
673/673 [==============================] - 165s 245ms/step - loss: 1.7356 - accuracy: 0.2942 - val_loss: 1.7370 - val_accuracy: 0.2916
Epoch 12/30
673/673 [==============================] - 180s 267ms/step - loss: 1.7258 - accuracy: 0.3003 - val_loss: 1.7340 - val_accuracy: 0.2997
Epoch 13/30
673/673 [==============================] - 172s 256ms/step - loss: 1.7204 - accuracy: 0.3063 - val_loss: 1.7190 - val_accuracy: 0.3061
Epoch 14/30
673/673 [==============================] - 134s 199ms/step - loss: 1.7148 - accuracy: 0.3085 - val_loss: 1.7215 - val_accuracy: 0.3049
Epoch 15/30
673/673 [==============================] - 160s 238ms/step - loss: 1.7091 - accuracy: 0.3091 - val_loss: 1.7166 - val_accuracy: 0.3128
Epoch 16/30
673/673 [==============================] - 191s 284ms/step - loss: 1.7052 - accuracy: 0.3145 - val_loss: 1.7117 - val_accuracy: 0.3104
Epoch 17/30
 81/673 [==>...........................] - ETA: 1:29 - loss: 1.7045 - accuracy: 0.3067

In [18]:
model.summary()

Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 swin_transformer_v2_tiny_wi  (None, 7, 7, 768)        27578154  
 ndow8 (Functional)                                              
                                                                 
 global_average_pooling2d (G  (None, 768)              0         
 lobalAveragePooling2D)                                          
                                                                 
 dropout (Dropout)           (None, 768)               0         
                                                                 
 dense (Dense)               (None, 7)                 5383      
                                                                 
Total params: 27,583,537
Trainable params: 27,583,537
Non-trainable params: 0
_________________________________________________________________


In [19]:
from keras.optimizers import Adam

In [22]:
model.compile(optimizer = Adam(1e-5) , loss = loss , metrics = ["accuracy"])

In [20]:
from tensorflow.keras.callbacks import EarlyStopping
early_stopping = EarlyStopping(
    monitor = "val_accuracy" , patience = 10 , restore_best_weights = True
)

In [21]:
from keras import callbacks

In [23]:
class_weights

{0: 1.02650269316936,
 1: 9.408038444735693,
 2: 1.0011157082422946,
 3: 0.568445781860416,
 4: 0.8261086389442995,
 5: 0.8491384409132132,
 6: 1.2931603915210472}

In [24]:

history = model.fit(train_generator, epochs = 50 , validation_data = Val_generator , callbacks = [early_stopping , Reduce_lr_plateau] , class_weight = class_weights)

Epoch 1/50
673/673 [==============================] - 2544s 4s/step - loss: 1.8691 - accuracy: 0.2328 - val_loss: 1.7202 - val_accuracy: 0.3577 - lr: 1.0000e-05
Epoch 2/50
673/673 [==============================] - 2244s 3s/step - loss: 1.6021 - accuracy: 0.4301 - val_loss: 1.4666 - val_accuracy: 0.5006 - lr: 1.0000e-05
Epoch 3/50
673/673 [==============================] - 2302s 3s/step - loss: 1.4481 - accuracy: 0.4977 - val_loss: 1.3979 - val_accuracy: 0.5275 - lr: 1.0000e-05
Epoch 4/50
673/673 [==============================] - 3110s 5s/step - loss: 1.3667 - accuracy: 0.5375 - val_loss: 1.3605 - val_accuracy: 0.5404 - lr: 1.0000e-05
Epoch 5/50
673/673 [==============================] - 8573s 13s/step - loss: 1.3054 - accuracy: 0.5592 - val_loss: 1.3487 - val_accuracy: 0.5531 - lr: 1.0000e-05
Epoch 6/50
673/673 [==============================] - 1836s 3s/step - loss: 1.2602 - accuracy: 0.5833 - val_loss: 1.2915 - val_accuracy: 0.5754 - lr: 1.0000e-05
Epoch 7/50
673/673 [=============

In [25]:
print("Swintiny with variable learning rate is finetuned fully. ")

Swintiny with variable learning rate is finetuned fully. 


## After finetuning 30 layers i am getting:
Epoch 1/30
673/673 [==============================] - 203s 295ms/step - loss: 1.4569 - accuracy: 0.4391 - val_loss: 1.3411 - val_accuracy: 0.4805
Epoch 2/30
673/673 [==============================] - 200s 297ms/step - loss: 1.2405 - accuracy: 0.5242 - val_loss: 1.2411 - val_accuracy: 0.5300
Epoch 3/30
673/673 [==============================] - 196s 291ms/step - loss: 1.1622 - accuracy: 0.5574 - val_loss: 1.2815 - val_accuracy: 0.5184
Epoch 4/30
673/673 [==============================] - 213s 316ms/step - loss: 1.1185 - accuracy: 0.5781 - val_loss: 1.1795 - val_accuracy: 0.5512
Epoch 5/30
673/673 [==============================] - 198s 294ms/step - loss: 1.0842 - accuracy: 0.5903 - val_loss: 1.2474 - val_accuracy: 0.5315
Epoch 6/30
673/673 [==============================] - 198s 294ms/step - loss: 1.0526 - accuracy: 0.6011 - val_loss: 1.1547 - val_accuracy: 0.5573
Epoch 7/30
673/673 [==============================] - 197s 293ms/step - loss: 1.0230 - accuracy: 0.6127 - val_loss: 1.2320 - val_accuracy: 0.5361
Epoch 8/30
673/673 [==============================] - 197s 292ms/step - loss: 0.9974 - accuracy: 0.6224 - val_loss: 1.0852 - val_accuracy: 0.5986
Epoch 9/30
673/673 [==============================] - 196s 291ms/step - loss: 0.9757 - accuracy: 0.6341 - val_loss: 1.0262 - val_accuracy: 0.6137
Epoch 10/30
673/673 [==============================] - 197s 292ms/step - loss: 0.9548 - accuracy: 0.6420 - val_loss: 1.0456 - val_accuracy: 0.6001
Epoch 11/30
673/673 [==============================] - 209s 310ms/step - loss: 0.9373 - accuracy: 0.6451 - val_loss: 1.0284 - val_accuracy: 0.6129
Epoch 12/30
673/673 [==============================] - 199s 296ms/step - loss: 0.9130 - accuracy: 0.6597 - val_loss: 1.0337 - val_accuracy: 0.6181
Epoch 13/30
673/673 [==============================] - 196s 291ms/step - loss: 0.8891 - accuracy: 0.6664 - val_loss: 1.0141 - val_accuracy: 0.6211
Epoch 14/30
 25/673 [>.............................] - ETA: 2:42 - loss: 0.7742 - accuracy: 0.7188

In [205]:
model.summary()

Model: "sequential_19"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 resnet50v2 (Functional)     (None, 7, 7, 2048)        23564800  
                                                                 
 global_average_pooling2d_12  (None, 2048)             0         
  (GlobalAveragePooling2D)                                       
                                                                 
 dense_41 (Dense)            (None, 7)                 14343     
                                                                 
Total params: 23,579,143
Trainable params: 14,343
Non-trainable params: 23,564,800
_________________________________________________________________


In [196]:
model = Sequential()
model.add(Conv2D(6 , kernel_size = (5 , 5) ,strides = (1 , 1) , padding = "valid",  activation =  "tanh" , input_shape = (224 , 224 , 3)))
model.add(AveragePooling2D(pool_size = (2 , 2) , strides = (2 , 2)))
model.add(Conv2D(16 , kernel_size = (5 , 5) , strides = (1 , 1) , padding = "valid" , activation = "tanh"))
model.add(AveragePooling2D(pool_size = (2 , 2) , strides = (2 , 2)))
model.add(Flatten())
model.add(Dense(120 , activation = "tanh"))
model.add(Dense(84 , activation = "tanh"))
model.add(Dense(7, activation = "softmax"))

Epoch 1/30
673/673 [==============================] - 139s 206ms/step - loss: 1.8339 - accuracy: 0.2379 - val_loss: 1.8131 - val_accuracy: 0.2513
Epoch 2/30
673/673 [==============================] - 170s 252ms/step - loss: 1.8198 - accuracy: 0.2456 - val_loss: 1.8129 - val_accuracy: 0.2513
Epoch 3/30
673/673 [==============================] - 150s 223ms/step - loss: 1.8174 - accuracy: 0.2484 - val_loss: 1.8174 - val_accuracy: 0.2513
Epoch 4/30
 39/673 [>.............................] - ETA: 1:34 - loss: 1.8053 - accuracy: 0.2532

## Moving towards ResNetV2

In [13]:
from tensorflow.keras.applications import ResNet50V2
from tensorflow.keras.applications.resnet50 import preprocess_input

In [14]:
from keras.layers import Dropout

In [99]:

base_model = ResNet50V2(input_shape = (224 , 224 , 3) , include_top= False , weights = 'imagenet')
base_model.trainable = True
model = Sequential(
    [
        base_model , GlobalAveragePooling2D() , Dense(512 , activation = "relu") , Dropout(0.5) ,  Dense(7 , activation = "softmax")
    ]
)

In [19]:
base_model.layers

In [100]:
for layer in base_model.layers[:-65]:
    layer.trainable = True

In [17]:
from sklearn.utils.class_weight import compute_class_weight
import numpy as np 
classes = np.unique(train_generator.classes)
weights = compute_class_weight(class_weight = 'balanced' , classes = classes , y = train_generator.classes)
class_weights = dict(zip(classes , weights))

In [18]:
weights

array([1.02650269, 9.40803844, 1.00111571, 0.56844578, 0.82610864,
       0.84913844, 1.29316039])

In [77]:
test_generator.reset()

In [27]:
pred = model.predict(test_generator)

225/225 [==============================] - 174s 764ms/step


In [28]:
for items in pred:
    print(items)

[0.8219627  0.00399158 0.09269688 0.01156785 0.01353582 0.02445207
 0.03179298]
[0.9718815  0.0011496  0.00698975 0.00569205 0.00229707 0.00666269
 0.00532706]
[0.04642436 0.00657441 0.21374565 0.01178056 0.04059238 0.67088795
 0.00999472]
[0.39180276 0.01414643 0.3300969  0.01771198 0.05739533 0.17775133
 0.01109536]
[0.7859774  0.0077988  0.09478872 0.0538871  0.01798634 0.01838857
 0.02117316]
[0.60603356 0.3313004  0.03058    0.00470225 0.0101812  0.00729457
 0.00990799]
[0.0856923  0.0247529  0.40124333 0.00743985 0.10323114 0.16550666
 0.21213385]
[0.8374517  0.00512367 0.02473472 0.07468142 0.01313313 0.02822676
 0.01664859]
[0.16706996 0.00514269 0.47888294 0.02798264 0.11737107 0.1866795
 0.01687121]
[0.88912076 0.00310731 0.07037832 0.01601265 0.00743254 0.00667384
 0.00727456]
[0.86932707 0.02296065 0.09319951 0.00363077 0.00396147 0.00492803
 0.0019925 ]
[0.43989047 0.11301205 0.08949996 0.05324331 0.15806687 0.12197119
 0.02431627]
[0.8743496  0.01116115 0.04907151 0.00766

In [47]:
pred1 = model.evalute(test_generator)

AttributeError: 'Sequential' object has no attribute 'evalute'

In [29]:
from sklearn.metrics import classification_report


In [30]:
import numpy as np

In [31]:
ans = np.argmax(pred , axis = 1)

In [32]:
Labels_test = test_generator.labels

In [33]:
len(Labels_test)

7178

In [34]:
len(ans)

7178

In [35]:
ans

array([0, 0, 5, ..., 6, 6, 6], dtype=int64)

In [36]:
classes = test_generator.classes

In [37]:
for items in classes:
    print(items)

0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0


In [38]:
print(classification_report(Labels_test , ans , zero_division = True))

              precision    recall  f1-score   support

           0       0.49      0.66      0.56       958
           1       0.58      0.61      0.59       111
           2       0.55      0.39      0.46      1024
           3       0.88      0.84      0.86      1774
           4       0.62      0.63      0.62      1233
           5       0.55      0.52      0.54      1247
           6       0.74      0.83      0.78       831

    accuracy                           0.66      7178
   macro avg       0.63      0.64      0.63      7178
weighted avg       0.66      0.66      0.65      7178



PERFORMANCE OF TRANSFER LEARNING ON EFFICIENTNETB3 WITH CLASS WEIGHTS: 
precision    recall  f1-score   support

           0       0.37      0.24      0.29       958
           1       0.17      0.29      0.21       111
           2       0.35      0.18      0.24      1024
           3       0.60      0.67      0.63      1774
           4       0.44      0.42      0.43      1233
           5       0.34      0.54      0.42      1247
           6       0.64      0.56      0.60       831

    accuracy                           0.46      7178
   macro avg       0.42      0.41      0.40      7178
weighted avg       0.46      0.46      0.45      7178

In [25]:
model.summary()

Model: "sequential_7"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 efficientnetb3 (Functional)  (None, 7, 7, 1536)       10783535  
                                                                 
 global_average_pooling2d_1   (None, 1536)             0         
 (GlobalAveragePooling2D)                                        
                                                                 
 dense_1 (Dense)             (None, 7)                 10759     
                                                                 
Total params: 10,794,294
Trainable params: 10,759
Non-trainable params: 10,783,535
_________________________________________________________________


In [39]:
model.save("Emotion detection model(Fully finetuned swin).keras" , model)

## Try next architecture. 1) EfficientNetB3 , 2) SwinTiny

In [16]:
from tensorflow.keras.applications import EfficientNetB3
from tensorflow.keras.applications.efficientnet import preprocess_input
from keras import Sequential
from keras.layers import Dropout

In [57]:
model = Sequential()
base_model = EfficientNetB3(weights = 'imagenet'  , input_shape = (224 , 224 , 3) , include_top = False)
base_model.trainable = True
model = Sequential(
    [
        base_model , GlobalAveragePooling2D() , Dense(512 , activation = "relu") , Dropout(0.5) ,   Dense(7 , activation = "softmax")
    ]
)

In [58]:
for layer in base_model.layers[:-45]:
    layer.trainable = True

## Results i am getting after finetuning 45 layers of EfficientNetB3 on this dataset. 
             precision    recall  f1-score   support

           0       0.61      0.49      0.55       958
           1       0.70      0.48      0.57       111
           2       0.47      0.45      0.46      1024
           3       0.91      0.81      0.86      1774
           4       0.61      0.60      0.61      1233
           5       0.48      0.63      0.55      1247
           6       0.72      0.82      0.77       831

    accuracy                           0.65      7178
   macro avg       0.64      0.61      0.62      7178
weighted avg       0.66      0.65      0.65      7178

## Results of Full finetuning of Swin Transformer with learning rate 1e-5 and l2 regularisation 1e-4.
precision    recall  f1-score   support

           0       0.55      0.54      0.54       958
           1       0.43      0.67      0.52       111
           2       0.50      0.43      0.46      1024
           3       0.90      0.81      0.85      1774
           4       0.55      0.68      0.61      1233
           5       0.54      0.49      0.52      1247
           6       0.71      0.82      0.76       831

    accuracy                           0.64      7178
   macro avg       0.60      0.63      0.61      7178
weighted avg       0.65      0.64      0.64      7178


## Now i will try SwinTiny 

In [21]:
import keras_cv_attention_models
help(keras_cv_attention_models)

Help on package keras_cv_attention_models:

NAME
    keras_cv_attention_models

PACKAGE CONTENTS
    aotnet (package)
    attention_layers (package)
    backend
    beit (package)
    botnet (package)
    caformer (package)
    clip (package)
    cmt (package)
    coat (package)
    coatnet (package)
    coco (package)
    common_layers
    convnext (package)
    cotnet (package)
    cspnext (package)
    davit (package)
    download_and_load
    edgenext (package)
    efficientdet (package)
    efficientformer (package)
    efficientnet (package)
    efficientvit (package)
    fasternet (package)
    fastervit (package)
    fastvit (package)
    gcvit (package)
    ghostnet (package)
    gpt2 (package)
    gpvit (package)
    halonet (package)
    hiera (package)
    hornet (package)
    iformer (package)
    imagenet (package)
    inceptionnext (package)
    keras_core_functional
    levit (package)
    llama2 (package)
    maxvit (package)
    mlp_family (package)
    mobilenetv3_fa

In [26]:
dir(swin_transformer_v2)

['ExpLogitScale',
 'MlpPairwisePositionalEmbedding',
 'SwinTransformerV2',
 'SwinTransformerV2Base_window12',
 'SwinTransformerV2Base_window16',
 'SwinTransformerV2Base_window24',
 'SwinTransformerV2Base_window8',
 'SwinTransformerV2Large_window12',
 'SwinTransformerV2Large_window16',
 'SwinTransformerV2Large_window24',
 'SwinTransformerV2Small_ns',
 'SwinTransformerV2Small_window16',
 'SwinTransformerV2Small_window8',
 'SwinTransformerV2Tiny_ns',
 'SwinTransformerV2Tiny_window16',
 'SwinTransformerV2Tiny_window8',
 'WindowAttentionMask',
 '__builtins__',
 '__cached__',
 '__default_doc__',
 '__doc__',
 '__file__',
 '__head_doc__',
 '__loader__',
 '__name__',
 '__package__',
 '__path__',
 '__spec__',
 '__tail_doc__',
 'shifted_window_attention',
 'swin_transformer_v2',
 'swin_transformer_v2_timm',
 'switch_to_deploy',
 'window_mhsa_with_pair_wise_positional_embedding']

In [9]:
from keras_cv_attention_models import swin_transformer_v2
base_model = swin_transformer_v2.SwinTransformerV2Tiny_window8(
    input_shape = (224 , 224 , 3) , num_classes = 0 , pretrained = "imagenet" 
)

c:\Users\wwwad\tf_directml310\lib\site-packages\tqdm\auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


>>>> Load pretrained from: C:\Users\wwwad\.keras\models\swin_transformer_v2_tiny_window8_256_imagenet.h5


In [10]:
from keras.layers import Dropout

In [11]:
model = Sequential(
    [
        base_model , GlobalAveragePooling2D() ,  Dropout(0.4) ,  Dense(7 , activation = "softmax" , kernel_regularizer = keras.regularizers.l2(1e-4))
    ]
)

In [15]:
base_model.summary()

Model: "swin_transformer_v2_tiny_window8"
__________________________________________________________________________________________________
 Layer (type)                   Output Shape         Param #     Connected to                     
 input_1 (InputLayer)           [(None, 224, 224, 3  0           []                               
                                )]                                                                
                                                                                                  
 stem_conv (Conv2D)             (None, 56, 56, 96)   4704        ['input_1[0][0]']                
                                                                                                  
 stem_ln (LayerNormalization)   (None, 56, 56, 96)   192         ['stem_conv[0][0]']              
                                                                                                  
 tf.reshape (TFOpLambda)        (None, 8, 7, 768)    0           ['

In [13]:
for layer in base_model.layers:
    layer.trainable = True

In [44]:
from sklearn.metrics import confusion_matrix
print(confusion_matrix(ans , Labels_test))

[[ 515   17  128   50   80  123   23]
 [  45   74   16    7    4   18    7]
 [ 101    4  441   37   75  166   57]
 [  24    3   17 1429   49   32   31]
 [ 103    6  125  148  835  270   20]
 [ 138    5  168   41  163  617   14]
 [  32    2  129   62   27   21  679]]


## Now applying some addition functionalities 

In [14]:
from keras.callbacks import ReduceLROnPlateau
from keras.losses import CategoricalCrossentropy
loss = CategoricalCrossentropy(label_smoothing = 0.1)
Reduce_lr_plateau = ReduceLROnPlateau(monitor = 'val_loss' , factor = 0.2 , patience = 3 , min_lr = 1e-7)

In [40]:
from tensorflow.keras.models import load_model

model = load_model(
    r"Emotion detection model(Fully finetuned swin).keras"
)

print(model.summary())

Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 swin_transformer_v2_tiny_wi  (None, 7, 7, 768)        27578154  
 ndow8 (Functional)                                              
                                                                 
 global_average_pooling2d (G  (None, 768)              0         
 lobalAveragePooling2D)                                          
                                                                 
 dropout (Dropout)           (None, 768)               0         
                                                                 
 dense (Dense)               (None, 7)                 5383      
                                                                 
Total params: 27,583,537
Trainable params: 27,583,537
Non-trainable params: 0
_________________________________________________________________
None


In [42]:
model.save_weights("emotion_weights.h5")